### DISCUSSION TITLE
Custom dynamic exports V1 migration

### DISCUSSION THREAD

**Discussion by chemeng** [vikejs/vike#1394] [Discussion] [answered]
Hi guys, I'm trying to migrate our codebase to V1 and I got almost everything working except our previous custom hooks.

As an example, here is our previous `index.page.server.js` file:
```
const serverRequest = {
    url: (ctx) => { return `addresses/localities/${ctx.routeParams.country}`},
    key: "localityUrls" 
};
export { serverRequest };
```

(the above is used in the global beforeRenderHook)

I tried using the approach described in the docs, creating a global and and a local `+config.h.js` file. This works well but I haven't been able to figure out the runtime code part.

If I do this in my `index/+config.h.js` file (after configuring meta in the global config file):
```
const serverRequest = {
    url: (ctx) => { return `addresses/localities/${ctx.routeParams.country}`},
    key: "localityUrls" 
};
export default {
  serverRequest: serverRequest
};
```
I get the warning that this dynamic code can't live inside this file. So I tried moving it elsewhere and then importing it like:

```
import { serverRequest } from "../../helpers/serverRequests.js";

export default { serverRequest: serverRequest};
```

but the code is not running (I get an "import" object at runtime). I'm assuming this happens because I'm not re-exporting the import? (https://vike.dev/header-file#not-re-exported-imports) . If I try re-exporting the import then I get an error that the serverRequest object is not included in `meta` in the global config file, etc.

I guess the question in the end boils down to implementing an example like this one: https://vike.dev/meta#example-sql but where the sql.js content is dynamic and depends on the context as well.

What's the right way to do this?

**Comment by lourot** [comment]
Hi @chemeng, from your message it sounds like you have understood the docs correctly and that you are doing things correctly, so it must be just a tiny detail that is wrong somewhere, but I couldn't spot it. Can you push your code somewhere so we can have a look?

**Reply by chemeng** [reply]
I was able to create a min repro here: https://stackblitz.com/edit/vitejs-vite-ynqv5o?file=src%2F%2BonBeforeRender.js

**Reply by lourot** [reply]
`yarn install` fails on Stackblitz:

```
~/projects/vitejs-vite-ynqv5o
❯ yarn install && yarn start
yarn install v1.22.19
warning package.json: "test" is also the name of a node core module
warning test@2.0.1: "test" is also the name of a node core module
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
error An unexpected error occurred: "too much recursion".
```

maybe it's running out of memory? Do you mind just pushing your code to a git repo that I can clone on my laptop? (I couldn't find a Download button on Stackblitz)

> [!NOTE]
> I'm mostly away from keyboard until Tuesday so bear with me if I don't quickly get back to you

**Reply by chemeng** [reply]
Very strange, it ran fine for me a few times I tried on stackblitz. Here is the code: https://github.com/chemeng/vike-debug-v1

**Reply by phonzammi** [reply]
Hi @chemeng,
How about [this](https://stackblitz.com/edit/vitejs-vite-nyb5jw?file=src%2Fhelpers%2FserverRequests.js) ?

**Reply by chemeng** [reply]
Ohh interesting, so the problem is defining the list within the `index/+config.h.js` (it's runtime execution I guess?)?

The problem with this is that in my application I have maybe 50 different objects and I import different combinations of them in the corresponding `+config.h.js` files for each page. I would have to do the "packaging" of the different objects into lists in a separate file and just import and assign (which is doable of course but introduces some extra code that feels unnecessary)

**Reply by phonzammi** [reply]
I'm sorry, I didn't realy understand your use case

**Reply by chemeng** [reply]
Let's say I have 10 data fetching requests that I've defined in my serverRequests.js as objects and I have 5 different pages I want to render.

Each page requires a subset of the 10 requests to be executed so I may have the following `+config.h.js` files:


```
// page1/+config.h.js
import {req1, req2} from serverRequests.js
export default {requests: [req1, req2]}
```

```
// page2/+config.h.js
import {req1, req3} from serverRequests.js
export default {requests: [req1, req3]}
```

```
// page3/+config.h.js
import {req2, req3} from serverRequests.js
export default {requests: [req2, req3]}
```

... etc

It looks like I would have to define all these arrays `[req1, req2]`, ..., `[req2, req3]` in a 3rd file and import them to the corresponding `+config.h.js` which adds an extra file per page.

**Reply by lourot** [reply]
Note that instead of using two files per page, you can use only one `+requests.js` per page. Because this:

```js
// page1/serverRequests.js
export const req1 = 'hello'
export const req2 = 'hi'

// page1/+config.h.js
import {req1, req2} from serverRequests.js
export default {requests: [req1, req2]}
```

is equivalent to:

```js
// page1/+requests.js
export default ['hello', 'hi']
```

You don't need a `+config.h.ts` file. See for example https://vike.dev/meta#example-sql and https://vike.dev/config#other-files

**Reply by phonzammi** [reply, marked as answer]
@chemeng, I've [updated](https://stackblitz.com/edit/vitejs-vite-nyb5jw?file=src%2Fhelpers%2FserverRequests.js,src%2F%2BonBeforeRender.js,src%2Fpages%2Findex%2F%2Brequests.js) as @AurelienLourot suggest.

**Reply by chemeng** [reply]
very nice, thanks guys!

**Reply by lourot** [reply]
@chemeng Glad we could help. If Vike works for you and you'd like to help, consider [contributing](https://github.com/vikejs/vike/issues/1349) or [sponsoring](https://github.com/sponsors/vikejs/). It makes a big difference :pray:
